sdk/nodejs/docker.ts

```diff
@@ -157,28 +157,31 @@ export function buildAndPushImageAsync(
 export function buildAndPushImage(
     imageName: string,
     pathOrBuild: pulumi.Input<string | DockerBuild>,
-    repositoryUrl: pulumi.Input<string>,
+    repositoryUrl: pulumi.Input<string> | undefined,
     logResource: pulumi.Resource,
     connectToRegistry?: () => pulumi.Input<Registry>,
     skipPush: boolean = false): pulumi.Output<string> {
 
-    return pulumi.all([pathOrBuild, repositoryUrl])
-        .apply(async ([pathOrBuildVal, repositoryUrlVal]) => {
-
-            // Give an initial message indicating what we're about to do.  That way, if anything
-            // takes a while, the user has an idea about what's going on.
-            logEphemeral("Starting docker build and push...", logResource);
+    // We do something rather interesting here.  We do not want to proceed if we don't actually have
+    // a value yet for `pathOrBuild`.  So we do a normal `ouput(...).apply(...)`.  However, we *do*
+    // want proceed if we don't have a value yet for `repositoryUrl`.  In that case, we'll just
+    // build without actually pushing.  To support that, we run `.apply` on the repoUrl, but we pass
+    // in `runWithUnknowns:true` to actually continue on in that case.
```

Comment:
note: higher level layers will likely need the same.  i.e. we call buildAndPushImage from inside Outputs higher up that will likely not run since they won't have a known repoUrl.  Will be making further changes in pulumi/awsx once this is in.